### PR TITLE
Uid concurrency check fix for vcards

### LINKED ISSUE #11387: Addressbooks should be check for uid conccurency before accepting a vcard
Like calendars: https://github.com/nextcloud/server/blob/db941a3f2e0f8e2afe61b972a808e0db9a82c876/apps/dav/lib/CalDAV/CalDavBackend.php#L978-L990

- [ ] New db column
- [ ] Migration path

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix #11387

- [x] Check
- [x] Tests

### CODE REVIEW

**Comment by skjnldsv** [comment]
For some reason the tests passes on my setup :)
Is it something like PostgreSQL sort database returns differently?

**Comment by MorrisJobke** [comment]
> Is it something like PostgreSQL sort database returns differently?

Ah yes ... there was something. @nickvergessen Do you know how we fixed that for Postgres? I can only remember that we did something with the sort order. 🤔

File: apps/dav/lib/CardDAV/CardDavBackend.php [hunk 1]
```diff
@@ -611,6 +611,18 @@ function createCard($addressBookId, $cardUri, $cardData) {
 		$etag = md5($cardData);
 		$uid = $this->getUID($cardData);
 
+		$q = $this->db->getQueryBuilder();
+		$q->select($q->createFunction('COUNT(*)'))
```
Comment (nickvergessen) [inline apps/dav/lib/CardDAV/CardDavBackend.php]: `$q->func()->count('*')`

Comment (nickvergessen) [reply apps/dav/lib/CardDAV/CardDavBackend.php]: But actually you dont want to count, so just select the uid
Also add `setMaxResults(1)` so the DB stops querying when it found an entry

Comment (skjnldsv) [reply apps/dav/lib/CardDAV/CardDavBackend.php]: Thanks! :)

Comment (skjnldsv) [reply apps/dav/lib/CardDAV/CardDavBackend.php]: Ah, I saw your second message too late.

**Review by MorrisJobke — APPROVED** [review]
Code makes sense 👍

**Comment by MorrisJobke** [comment]
@skjnldsv Do we want to backport this? Or is this column only available in master anyways?

**Comment by skjnldsv** [comment]
> @skjnldsv Do we want to backport this? Or is this column only available in master anyways?

master only.
So no backport :(

**Comment by skjnldsv** [comment]
Thanks a lot @nickvergessen! :hugs:

**Comment by skjnldsv** [comment]
`sh: 1: kill: No such process` unrelated
And tags acceptance is failing everywhere as well: unrelated.

**Comment by MorrisJobke** [comment]
@nickvergessen Just say here if a revert is needed.

**Comment by nickvergessen** [comment]
:heavy_check_mark: approved

**Pull request merged**